src/index.ts

```diff
@@ -154,37 +154,75 @@ export function WithFinalAuthTo<
 // @todo: implement all the other HTPipe*'s - note that each one will be slightly different based on their specifics...
 // i.e. can return bool vs not, possibly async vs sync only, mandatory vs not mandatory...
 // then the final master HTPipe will just build an object out of all the sub-HTPipe*'s
+
+// Left and right have attach data
 export function HTPipeAttachData<
-  TLeft extends OptionallyHasAttachData<
+  TLeft extends HasAttachData<
+    any,
+    Partial<Parameters<TRight['attachData']>[0]>
+  >,
+  TRight extends HasAttachData<any, any>,
+  TContextInLeft extends Parameters<TLeft['attachData']>[0],
+  TContextInRight extends Parameters<TRight['attachData']>[0],
+  TContextOutLeft extends PromiseResolveOrSync<ReturnType<TLeft['attachData']>>,
+  TContextOutRight extends PromiseResolveOrSync<
+    ReturnType<TRight['attachData']>
+  >
+>(
+  left: TLeft,
+  right: TRight
+): HasAttachData<
+  TContextInLeft & Omit<TContextInRight, keyof TContextOutLeft>,
+  TContextOutLeft & TContextOutRight
+>;
+
+// only left has attach data
+export function HTPipeAttachData<
+  TLeft extends HasAttachData<
     any,
     TRight extends HasAttachData<any, any>
       ? Partial<Parameters<TRight['attachData']>[0]>
       : any
   >,
   TRight extends OptionallyHasAttachData<any, any>,
-  TContextInLeft extends TLeft extends HasAttachData<any, any>
-    ? Parameters<TLeft['attachData']>[0]
-    : {},
-  TContextInRight extends TRight extends HasAttachData<any, any>
-    ? Parameters<TRight['attachData']>[0]
-    : {},
-  TContextOutLeft extends TLeft extends HasAttachData<any, any>
-    ? PromiseResolveOrSync<ReturnType<TLeft['attachData']>>
-    : {},
-  TContextOutRight extends TRight extends HasAttachData<any, any>
-    ? PromiseResolveOrSync<ReturnType<TRight['attachData']>>
-    : {}
+  TContextInLeft extends Parameters<TLeft['attachData']>[0],
+  TContextOutLeft extends PromiseResolveOrSync<ReturnType<TLeft['attachData']>>
 >(
   left: TLeft,
   right: TRight
 ): HasAttachData<
-  TContextInLeft & Omit<TContextInRight, keyof TContextOutLeft>,
-  TContextOutLeft & TContextOutRight
```

Comment:
I get that you're trying to keep it structurally similar to the other overloads, but I think in this case they are so far apart and it's so complicated that it's better to keep it simpler (i.e. remove ` & Omit<{}, keyof TContextOutLeft>,` and `& {}`

Reply:
likewise with the right-only
- [ ] clarity over consistency x2

Reply:
done